API: завершить смену

DD477 opened the pull request.

DD477 committed:
added ability to end shifts

DD477 committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into feature/api_shift_finish

FinemechanicPub commented on src/bot/services.py:
```diff
@@ -73,3 +73,7 @@ async def notify_excluded_member(self, user: models.User) -> None:
+
+    async def notify_that_shift_is_finished(self, user: models.User, shift: models.Shift) -> None:
+        """Уведомляет участников об окончании."""
+        await self.__bot.send_message(user.telegram_id, shift.final_message)
```
В final message нужно подставить имя, фамилию участника и количество ломбарьерчиков. Формат сообщения можно посмотреть в shift_service.py

DD477 committed:
added mass mailing

AntonZelinsky commented on src/core/services/shift_service.py:
```diff
@@ -61,6 +63,16 @@ async def start_shift(self, id: UUID) -> Shift:
+    async def finish_shift(self, bot, id: UUID) -> Shift:
+        shift = await self.__shift_repository.get(id)
+        if shift.status == Shift.Status.FINISHED.value:
+            raise NotFoundException(object_name=Shift.__doc__, object_id=id)
```
СОздай ещё одно исключение, тут ты нашёл смену, но не можешь её финишировать. 
Кстати, ты не можешь финишировать не только уже финишированную смену, но и PREPARING и CANCELING. Делай проверку, если смена не STARTED, то её нельзя финишировать.

AntonZelinsky commented on src/core/services/shift_service.py:
```diff
@@ -61,6 +63,16 @@ async def start_shift(self, id: UUID) -> Shift:
+        if shift.status == Shift.Status.FINISHED.value:
+            raise NotFoundException(object_name=Shift.__doc__, object_id=id)
+        update_shift_dict = {"finished_at": datetime.now().date(), "status": Shift.Status.FINISHED.value}
+        updated_shift = await self.__shift_repository.update(id=id, instance=Shift(**update_shift_dict))
```
Так нельзя делать!  
@ADesBiysk я где-то писал почему, я даже просил сделать пост кого-то из студентов, если увидишь, то пни чтобы создали пост почему нельзя.


В этом методе ты получаешь shift, вот прямо в него и устанавливай значения, а потом передавай в апдейт.

AntonZelinsky replied:
Заодно # noqa: R504 уйдёт. Хотя её и так не должно быть. странно.

AntonZelinsky replied:
[Тут описывал как сделать по DDD,](https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/pull/110#discussion_r1034176233), советую и тут так же сделать, просто метод у модели смены - `shift.finish_shift()`

DD477 replied:
Не понял последнее.

AntonZelinsky replied:
почитай что по ссылке писал.
Сделать метод finish_shift у модели Shift, внутри этого метода проверяй свойство статус на то, можно ли смену финишировать и если нет, то возвращай исключение, если всё ок, то устанавливай присваиванием новое значение статусу и дате финиширования и выходи и функции. А в сервисе вызывай метод репозитория сохранения и передавай модель туда

DD477 replied:
Добавил.

FinemechanicPub replied:
>Сделать метод finish_shift у модели Shift, внутри этого метода проверяй свойство статус на то, можно ли смену финишировать и если нет, то возвращай исключение, если всё ок, то устанавливай присваиванием новое значение статусу и дате финиширования и выходи и функции.

Проверка статуса? То есть бизнес-логика съехала уже даже не в репозиторий, а в модель. Я перестал понимать архитектуру этого проекта.

mspiridonov2706 replied:
@AntonZelinsky могу по такой же логике исправить start_shift? И исключение NotFoundException в start_shift изменить на ShiftStartImpossibleException? Заодно исправлю баг с update во время вызова функции start_shift. Выбрасывает исключение во время обновления.

DD477 committed:
Fix shift update

DD477 committed:
fixed after review

DD477 committed:
added finish_shift method in shift model

FinemechanicPub commented on src/core/db/models.py:
```diff
@@ -59,6 +62,11 @@ class Status(str, enum.Enum):
     def __repr__(self):
         return f"<Shift: {self.id}, status: {self.status}>"
 
+    def finish_shift(self, shift):
```
Зачем параметр shift? Мы же хотим завершить эту, то есть, self смену.

FinemechanicPub commented on src/core/db/models.py:
```diff
@@ -59,6 +62,11 @@ class Status(str, enum.Enum):
+    def finish_shift(self, shift):
+        if shift.status != Shift.Status.STARTED.value:
+            raise NotFoundException(object_name=self.__doc__, object_id=id)
+        return {"finished_at": datetime.now().date(), "status": self.Status.FINISHED.value}
```
Я думаю, этот метод не должен ничего возвращать. Это метод смены, который эту же смену отмечает завершенной. Просто объект смены, у которого метод вызван, должен становиться FINISHED с соответствующей датой окончания. Нужно модифицировать себя (self) и всё.

AntonZelinsky replied:
И дату в self.finished-at устанавливать в этом методе

FinemechanicPub commented on src/core/services/shift_service.py:
```diff
@@ -61,6 +63,15 @@ async def start_shift(self, id: UUID) -> Shift:
 
+    async def finish_shift(self, bot, id: UUID) -> Shift:
+        shift = await self.__shift_repository.get(id)
+        update_shift_dict = shift.finish_shift(shift=shift)
```
А здесь просто shift.finish_shift()

AntonZelinsky replied:
Давайте переименуем на просто финиш, а то тавтология не красивая

FinemechanicPub commented on src/core/services/shift_service.py:
```diff
@@ -61,6 +63,15 @@ async def start_shift(self, id: UUID) -> Shift:
+    async def finish_shift(self, bot, id: UUID) -> Shift:
+        shift = await self.__shift_repository.get(id)
+        update_shift_dict = shift.finish_shift(shift=shift)
+        await self.__shift_repository.update(id=id, instance=Shift(**update_shift_dict))
```
А update вызывать не для собранного из словаря объекта Shift, а того, который чуть выше уже модифицировали.

AntonZelinsky commented on src/core/db/models.py:
```diff
@@ -59,6 +62,11 @@ class Status(str, enum.Enum):
 
+    def finish_shift(self, shift):
+        if shift.status != Shift.Status.STARTED.value:
+            raise NotFoundException(object_name=self.__doc__, object_id=id)
```
Неверное исключение. Тут ошибка смена не запущена, а не "не найден". Создай исключение ШифтНотСтартедЭррор.

AntonZelinsky commented on src/core/services/shift_service.py:
```diff
@@ -61,6 +63,15 @@ async def start_shift(self, id: UUID) -> Shift:
+        shift = await self.__shift_repository.get(id)
+        update_shift_dict = shift.finish_shift(shift=shift)
+        await self.__shift_repository.update(id=id, instance=Shift(**update_shift_dict))
+        updated_shift = await self.__shift_repository.get_with_users(id=id)
```
Просто передавай id, не нужно id=id

AntonZelinsky commented on src/core/services/shift_service.py:
```diff
@@ -61,6 +63,15 @@ async def start_shift(self, id: UUID) -> Shift:
+        await self.__shift_repository.update(id=id, instance=Shift(**update_shift_dict))
+        updated_shift = await self.__shift_repository.get_with_users(id=id)
+        for user in updated_shift.users:
+            await self.__telegram_bot(bot).notify_that_shift_is_finished(user)
```
Прям плохо, что так создаётся бот сервис. Посмотри как в других местах сделано, должны быть варианты получше. Если нет, то придётся придумать.

mspiridonov2706 replied:
Антон, добрый вечер. В других сервисах также сделано([тут](https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/blob/472125f5b3fe63745c27fc5e728d5660ec4acf06/src/core/services/user_task_service.py#L98) и [тут](https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/blob/472125f5b3fe63745c27fc5e728d5660ec4acf06/src/core/services/request_sevice.py#L26) например).

Оставить как есть? Или нужно переделать по-другому везде или только тут?

AntonZelinsky replied:
Блин, оставляй как есть..( обсудим и в другом МРе поправим

Commit:
Merge branch 'develop' into feature/api_shift_finish

AntonZelinsky commented on src/bot/services.py:
```diff
@@ -73,3 +73,13 @@ async def notify_excluded_member(self, user: models.User) -> None:
+
+    async def notify_that_shift_is_finished(self, user: models.User) -> None:
+        """Уведомляет участников об окончании смены."""
+        final_message = (
+            "Привет, {user.name} {user.surname}!"
```
Во [первых](https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/blob/472125f5b3fe63745c27fc5e728d5660ec4acf06/src/core/services/shift_service.py#L26), во [вторых](https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/blob/472125f5b3fe63745c27fc5e728d5660ec4acf06/src/core/services/shift_service.py#L46)! То есть, это сообщение уже есть в самой смене. Нужно от туда брать.

Commit:
fix: review

fix message format;
add new exceptions;
add start method in Shift model;
change start logic;
fix bug with primary_key in sequence_number;

mspiridonov2706 commented on src/core/db/models.py:
```diff
@@ -50,7 +56,7 @@ class Status(str, enum.Enum):
         Enum(Status, name="shift_status", values_callable=lambda obj: [e.value for e in obj]), nullable=False
     )
-    sequence_number = Column(Integer, Identity(start=1, cycle=True), primary_key=True)
+    sequence_number = Column(Integer, Identity(start=1, cycle=True))
```
Из-за этого primary_key был баг с методом update. Срабатывало исключение при обновлении данных. В миграциях это никак не учитывалось, поэтому просто удалил из описания модели и всё стало работать нормально. Решил, что по ошибке добавлено.

Вот [тут](https://practicum-students.slack.com/archives/C03SXFKDWJC/p1669931143105709) с Александром обсуждали этот баг.

AntonZelinsky replied:
Да, тут действительно не нужен первичный ключ, так как это не внешний ключ, а просто счётчик. Да и два первичных ключа быть не может

mspiridonov2706 commented on src/core/services/shift_service.py:
```diff
@@ -59,19 +59,26 @@ async def get_shift(self, id: UUID) -> Shift:
+        shift = Shift(**update_shift_data.dict(exclude_unset=True))
+        return await self.__shift_repository.update(id, shift)
 
     async def start_shift(self, id: UUID) -> Shift:
```
Поправил функцию start_shift по аналогии с finish_shift

Commit:
refactor: update_shift func refactoring

mspiridonov2706 commented on src/core/services/shift_service.py:
```diff
@@ -59,19 +59,25 @@ async def get_shift(self, id: UUID) -> Shift:
-        return await self.__shift_repository.update(id=id, instance=Shift(**update_shift_data.dict(exclude_unset=True)))
+        return await self.__shift_repository.update(id, Shift(**update_shift_data.dict(exclude_unset=True)))
 
     async def start_shift(self, id: UUID) -> Shift:
```
Поправил функцию start_shift по аналогии с finish_shift

Commit:
Merge branch 'develop' into feature/api_shift_finish

AntonZelinsky commented on src/core/services/shift_service.py:
```diff
@@ -16,16 +15,16 @@
 from src.core.services.task_service import TaskService
 from src.core.services.user_task_service import UserTaskService
 
 FINAL_MESSAGE = (
```
Эта штука больше не используется, удали её

ADesBiysk replied:
FINAL_MESSAGE используется в функции create_new_shift при создании смены, как сообщение по умолчанию.

AntonZelinsky commented on src/core/services/shift_service.py:
```diff
@@ -59,19 +59,25 @@ async def get_shift(self, id: UUID) -> Shift:
+        await shift.finish()
+        await self.__shift_repository.update(id, shift)
+        for user in shift.users:
+            await self.__telegram_bot(bot).notify_that_shift_is_finished(user, shift.final_message)
```
А как насчёт того, чтобы в функцию передавать только смену, а внутри уже циклом проходиться по участникам.
Кстати, ты проходишься не по участникам (members), а по юзерам, там есть какая-то разница, к тому же, в каком-то другом ПРе эту связь уже удалили, оставили только мемберс. И ещё один момент, и в одном и в другом случае будут приходить все участники, а нам нужно отправлять только тем, кто активный, не беря в расчёт исключённых.
Я могу быть не прав, лучше проверить. Как вариант, смержить этот ПР, а в другом внести правки. 
@ADesBiysk

AntonZelinsky replied:
И если нужно сделать много маленьких действий не требующих этого последовательно, то и не нужно делать их последовательно. Нужно запустить все задачи, дождаться выполнения их всех

ADesBiysk replied:
Лучше отдельным PR. Завёл задачу.

AntonZelinsky commented on src/core/services/shift_service.py:
```diff
@@ -59,19 +59,25 @@ async def get_shift(self, id: UUID) -> Shift:
 
     async def update_shift(self, id: UUID, update_shift_data: ShiftUpdateRequest) -> Shift:
-        return await self.__shift_repository.update(id=id, instance=Shift(**update_shift_data.dict(exclude_unset=True)))
+        return await self.__shift_repository.update(id, Shift(**update_shift_data.dict(exclude_unset=True)))
```
Уже много раз говорл, не использовать такое обновление данных..( 
Вообще, это очень пробленая функция, поскольку в ней нет никаких валидаций, кроме входных данных.
Если смена запустилась, нельзя редактировать дату старта, если финишировала, то по хорошему, нельзя редактировать ничего кроме названия. У запущенной смены нельзя дату финиша сделать предыдущим числом. Много ещё чего можно придумать. 
Сами модели запросов кривые.
![image](https://user-images.githubusercontent.com/1431279/205449500-5f0e4626-dca9-4f4e-8814-99e1cfafd061.png)
@ADesBiysk, можно тоже в отдельной задачке пустить, чтобы эту побыстрее смержить. Но задачи важно позоводить сейчас, чтобы не забыть

ADesBiysk replied:
Завёл задачу.

mspiridonov2706 replied:
Согласен, мне тоже такое обновление не очень нравится. Решил оставить из-за того, что у нас функция merge используется при обновлении.

ADesBiysk commented on src/core/services/shift_service.py:
```diff
@@ -59,19 +59,25 @@ async def get_shift(self, id: UUID) -> Shift:
-            raise NotFoundException(object_name=Shift.__doc__, object_id=id)
+        await shift.start()
 
         # TODO добавить вызов метода рассылки участникам первого задания
```
Комментарий уже не актуален.

mspiridonov2706 replied:
Поправил

AntonZelinsky commented on src/core/db/models.py:
```diff
@@ -197,4 +213,4 @@ def send_report(self, photo_url: str):
         self.status = UserTask.Status.UNDER_REVIEW.value
         self.report_url = photo_url
-        self.uploaded_at = datetime.datetime.now()
+        self.uploaded_at = datetime.now()
```
Вообще, по-хорошему, везде нужно писать время в базу в UTC, а уже при выводе пользователю приводить к его таймзоне.
Но это точно отдельная задача. Нужно везде делать одинаково.

AntonZelinsky commented on src/core/exceptions.py:
```diff
@@ -36,3 +36,15 @@ class DuplicateReportError(Exception):
+class ShiftStartForbiddenException(ApplicationException):
+    def __init__(self, object_name: str, object_id: UUID):
+        self.status_code = HTTPStatus.BAD_REQUEST
+        self.detail = f"Невозможно начать смену {object_name} с id: {object_id}. Проверьте статус смены"
```
Тут же ты даёшь очень узко конкретное название исключению, зачем тогда абстрактное название переменным инициализатора давать?

mspiridonov2706 replied:
Хотел, чтобы во всех исключениях одинаковые названия переменных были.

mspiridonov2706 replied:
Поправил

Commit:
fix: change variable names in Shift exceptions

Commit:
fix: variables names in raise exceptions shift

Commit:
fix: delete unnecessary comment
